src/workflow/output_validator.ts

```diff
@@ -1,75 +1,73 @@
-      if (parentName !== workflowName) {
-        throw new WorkflowCreationError(
-          `Invalid output reference: ${match[0]}, step is not a child of ${parentName}`,
-          this.workflow.name,
-          parentName,
-          stepName,
-        );
-      }
-      const outputName = `${workflowName}.${match[2]}`;
-      // The referenced child step output is already executed.
-      if (!this.seenSteps.has(outputName)) {
-        throw new WorkflowCreationError(
-          `Invalid output reference: ${match[0]}`,
-          this.workflow.name,
-          parentName,
-          stepName,
-        );
-      }
+  validateWorkflowOutputReference(match: RegExpMatchArray, stepName: string, parentName?: string) {
+    const workflowName = match[2]; // The name of the workflow step
+    // Access to the child step outputs is restricted to within the same parent workflow step;
+    if (parentName !== workflowName) {
+      throw new WorkflowCreationError(
+        `Invalid output reference: ${match[0]}, step is not a child of ${parentName}`,
+        this.workflow.name,
+        parentName,
+        stepName,
+      );
     }
   }
 
-  validateSimpleOutputReferences(stepName: string, template: string, parentName?: string) {
-    const simpleOutputMatches = [...template.matchAll(regexSimpleOutput)];
-    // Multiple outputs may exist within the template so we need a loop.
-    // In this case, we are looking for simple step output references.
-    // Format: $.outputs.stepName
-    // Example template: $.outputs.stepName1 + " " + $.outputs.stepName2
-    for (const match of simpleOutputMatches) {
-      const outputStepName = match[1];
-      // The referenced step output is already executed.
-      if (!this.seenSteps.has(outputStepName)) {
-        throw new WorkflowCreationError(
-          `Invalid output reference: ${match[0]}`,
-          this.workflow.name,
-          parentName ?? stepName,
-          parentName ? stepName : undefined,
-        );
-      }
+  validateExistanceOfOutputReference(
+    match: RegExpMatchArray,
+    stepName: string,
+    parentName?: string,
+  ) {
+    const fullOutputName = match[1]; // For workflow step
+    const outputStepName = match[2]; // For simple step
+    // Check the referenced step output is already executed.
+    if (!this.seenSteps.has(fullOutputName) && !this.seenSteps.has(outputStepName)) {
+      throw new WorkflowCreationError(
+        `Invalid output reference: ${match[0]}, step is not executed yet.`,
+        this.workflow.name,
+        parentName ?? stepName,
+        parentName ? stepName : undefined,
+      );
     }
   }
 
   validateOutputReferences(stepName: string, template?: string, parentName?: string) {
     if (!template) {
       return;
     }
-    this.validateWorkflowOutputReferences(template, stepName, parentName);
-    this.validateSimpleOutputReferences(stepName, template, parentName);
+    const outputMatches = [...template.matchAll(regexOutputReference)];
```

Comment:
[minor] Do we need destructuring here ? I think `matchAll` already gives us array

Reply:
yes we need it as matchAll gives an iterator, not an array so restructuring makes it array

Reply:
Got it 👍🏼